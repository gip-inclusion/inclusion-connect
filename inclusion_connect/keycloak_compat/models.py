from django.db import models


class JWTHashSecret(models.Model):
    realm_id = models.TextField(primary_key=True)
    # urlsafe_base64_encoded.
    secret = models.TextField()
