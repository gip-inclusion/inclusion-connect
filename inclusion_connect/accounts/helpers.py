from django.contrib import auth


def login(request, user):
    """
    Log the user and preserve the next url (as login again flushes the session)
    """
    next_url = request.session.get("next_url")
    auth.login(request, user)
    if next_url:
        request.session["next_url"] = next_url
