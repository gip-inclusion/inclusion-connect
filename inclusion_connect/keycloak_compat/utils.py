import re


namespace_pattern = re.compile(r"^keycloak_compat_(?P<realm>.+)$")


def realm_from_request(request):
    namespace = request.resolver_match.namespace
    match = namespace_pattern.match(namespace)
    return match.group("realm")
