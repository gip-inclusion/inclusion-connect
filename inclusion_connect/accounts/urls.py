from django.urls import path, re_path

from inclusion_connect.accounts import views


app_name = "accounts"

urlpatterns = [
    re_path(r"^login/$", views.LoginView.as_view(), name="login"),
    re_path(r"^register/$", views.RegisterView.as_view(), name="register"),
    re_path(r"^activate/$", views.ActivateAccountView.as_view(), name="activate"),
    re_path(r"^password_reset/$", views.PasswordResetView.as_view(), name="password_reset"),
    path("reset/<uidb64>/<token>/", views.PasswordResetConfirmView.as_view(), name="password_reset_confirm"),
    re_path(r"^accept-terms/$", views.AcceptTermsView.as_view(), name="accept_terms"),
    re_path(r"^my-account/$", views.EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^change-password/$", views.PasswordChangeView.as_view(), name="change_password"),
]
