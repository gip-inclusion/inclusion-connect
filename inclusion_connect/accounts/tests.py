from urllib.parse import quote

from django.contrib import messages
from django.contrib.auth import get_user
from django.core import mail
from django.urls import reverse
from django.utils.encoding import force_bytes
from django.utils.http import urlsafe_base64_encode
from pytest_django.asserts import assertContains, assertRedirects, assertTemplateUsed

from inclusion_connect.accounts.views import PasswordResetView
from inclusion_connect.users.factories import DEFAULT_PASSWORD, UserFactory
from inclusion_connect.users.models import User
from inclusion_connect.utils.urls import add_url_params


def test_login(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Connexion")
    assertContains(response, "Adresse e-mail")  # Ask for email, not username
    assertContains(response, reverse("accounts:registration"))  # Link to registration page

    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated


def test_login_failed_bad_email_or_password(client):
    url = add_url_params(reverse("accounts:login"), {"next": "anything"})
    user = UserFactory()
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": user.email, "password": "toto"})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    response = client.post(url, data={"email": "wrong@email.com", "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated

    # If user is inactive
    user.is_active = False
    user.save()
    response = client.post(url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assertTemplateUsed("login.html")
    assertContains(response, "Adresse e-mail ou mot de passe invalide.")
    assert not get_user(client).is_authenticated


def test_user_creation(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.get(url)
    assertContains(response, "Créer un compte")
    assertContains(response, reverse("accounts:login"))  # Link to login page
    assertContains(response, "CGU_20230302.pdf")
    assertContains(response, quote("Politique de confidentialité_20230302.pdf"))

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
            "terms_accepted": True,
        },
    )
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
    user = User.objects.get(email=user.email)  # Previous instance was a built factory, so refresh_from_db won't work
    assert user.terms_accepted_at == user.date_joined


def test_user_creation_terms_are_required(client):
    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:registration"), {"next": redirect_url})
    user = UserFactory.build()
    assert not get_user(client).is_authenticated

    response = client.post(
        url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
        },
    )
    assertTemplateUsed("registration.html")
    assert "terms_accepted" in response.context["form"].errors


# TODO: Test next url propagation in all the registration process (creation + email validation)


def test_password_reset(client):
    user = UserFactory()

    redirect_url = reverse("oidc_overrides:logout")
    url = add_url_params(reverse("accounts:login"), {"next": redirect_url})
    response = client.get(url)
    password_reset_url = reverse("accounts:password_reset")
    assertContains(response, password_reset_url)

    response = client.get(password_reset_url)
    assertTemplateUsed(response, "password_reset.html")

    response = client.post(password_reset_url, data={"email": user.email})
    assertRedirects(response, reverse("accounts:login"))
    assert list(messages.get_messages(response.wsgi_request)) == [
        messages.storage.base.Message(
            messages.SUCCESS,
            "Si un compte existe avec cette adresse e-mail, "
            "vous recevrez un e-mail contenant des instructions pour réinitialiser votre mot de passe.",
        ),
    ]

    # Check sent email
    assert len(mail.outbox) == 1
    uid = urlsafe_base64_encode(force_bytes(user.pk))
    token = PasswordResetView.token_generator.make_token(user)
    password_reset_url = reverse("accounts:password_reset_confirm", args=(uid, token))
    assert password_reset_url in mail.outbox[0].body

    # Change password
    password = "toto"
    response = client.get(password_reset_url)  # retrieve the modified url
    response = client.post(response.url, data={"new_password1": password, "new_password2": password})
    # FIXME: skip redirection to login, go to next url !
    assertRedirects(response, redirect_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated
