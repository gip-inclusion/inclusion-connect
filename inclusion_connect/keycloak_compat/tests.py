import uuid

import jwt
import pytest
from django.contrib.auth import get_user
from django.contrib.auth.hashers import PBKDF2PasswordHasher
from django.urls import reverse
from oauth2_provider.models import get_access_token_model, get_id_token_model, get_refresh_token_model
from pytest_django.asserts import assertRedirects

from inclusion_connect.keycloak_compat.hashers import KeycloakPasswordHasher
from inclusion_connect.oidc_overrides.factories import DEFAULT_CLIENT_SECRET, ApplicationFactory, default_client_secret
from inclusion_connect.users.factories import UserFactory
from inclusion_connect.utils.urls import add_url_params, get_url_params


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_login(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test AUTH endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:authorize")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))
    assert client.session["next_url"] == auth_complete_url

    # Test AUTH endpoint when not authenticated and with bad params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test AUTH endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


@pytest.mark.parametrize("realm", ["local", "Review_apps", "Demo", "inclusion-connect"])
def test_registration(client, realm):
    application = ApplicationFactory()
    user = UserFactory()

    # Test REGISTRATIONS endpoint when not authenticated
    auth_url = reverse(f"keycloak_compat_{realm}:registrations")
    auth_params = {
        "response_type": "code",
        "client_id": application.client_id,
        "redirect_uri": "http://localhost/callback",
        "scope": "openid profile email",
        "state": "state",
        "nonce": "nonce",
    }
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:registration"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when not authenticated and with bad oidc params
    bad_auth_params = auth_params.copy()
    bad_auth_params["client_id"] = "toto"
    bad_auth_complete_url = add_url_params(auth_url, bad_auth_params)
    response = client.get(bad_auth_complete_url)
    # FIXME update the template
    assert response.status_code == 400

    # Test REGISTRATIONS endpoint when not authenticated with activation params
    activation_params = auth_params | {
        "login_hint": "email",
        "firstname": "John",
        "lastname": "Doe",
    }
    auth_complete_url = add_url_params(auth_url, activation_params)
    response = client.get(auth_complete_url, follow=True)
    assertRedirects(response, reverse("accounts:activation"))
    assert client.session["next_url"] == auth_complete_url

    # Test REGISTRATIONS endpoint when authenticated
    client.force_login(user)
    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(auth_params["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    # Test TOKEN endpoint
    # FIXME it's recommanded to use basic auth here, maybe update our documentation ?
    token_data = {
        "client_id": application.client_id,
        "client_secret": DEFAULT_CLIENT_SECRET,
        "code": auth_response_params["code"],
        "grant_type": "authorization_code",
        "redirect_uri": "http://localhost/callback",
    }
    response = client.post(
        reverse(f"keycloak_compat_{realm}:token"),
        data=token_data,
    )
    token_json = response.json()
    id_token = token_json["id_token"]
    decoded_id_token = jwt.decode(
        id_token,
        key=default_client_secret(),
        algorithms=["HS256"],
        audience=application.client_id,
    )
    assert decoded_id_token["nonce"] == auth_params["nonce"]
    assert decoded_id_token["sub"] == str(user.pk)
    assert uuid.UUID(decoded_id_token["sub"]), "Sub should be an uuid"
    assert decoded_id_token["given_name"] == user.first_name
    assert decoded_id_token["family_name"] == user.last_name
    assert decoded_id_token["email"] == user.email

    # Test USER INFO endpoint
    response = client.get(
        reverse(f"keycloak_compat_{realm}:user-info"),
        HTTP_AUTHORIZATION=f"Bearer {token_json['access_token']}",
    )
    assert response.json() == {
        "sub": str(user.pk),
        "given_name": user.first_name,
        "family_name": user.last_name,
        "email": user.email,
    }

    # Test LOGOUT endpoint
    assert get_user(client).is_authenticated
    logout_params = {"id_token_hint": id_token}
    response = client.get(add_url_params(reverse(f"keycloak_compat_{realm}:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert get_id_token_model().objects.count() == 0
    assert get_access_token_model().objects.count() == 0
    assert get_refresh_token_model().objects.get().revoked is not None


def test_password_hasher(client):
    password = "RdaRfqP7Y89vy2"
    hashed_password = "$".join(
        [
            "keycloak-pbkdf2-sha256",
            "27500",
            "Td6XuopYK6JNfUnIlqYMOQ==",
            "ZXVC08Hf4jBOoYzVoNWYjQijsMC2oc/OUa9LciiIJ/1XHPF/qPiY1DqwLLDN2hYFmf/1kApkveD8/Pr7GVqjgw==",
        ]
    )
    user = UserFactory(password=hashed_password)
    assert user.password == hashed_password

    assert KeycloakPasswordHasher().verify(password=password, encoded=hashed_password)

    client.login(email=user.email, password=password)

    user.refresh_from_db()
    assert user.password != hashed_password
    assert PBKDF2PasswordHasher().verify(password, encoded=user.password)
