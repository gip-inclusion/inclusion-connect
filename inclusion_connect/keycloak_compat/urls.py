from django.urls import re_path
from oauth2_provider import views as oauth2_views

from ..accounts.views import EditUserInfoView
from ..oidc_overrides.views import AuthorizationView, LogoutView, RegistrationView
from . import views


app_name = "keycloak_compat"


urlpatterns = [
    re_path(
        r"^\.well-known/openid-configuration/$",
        oauth2_views.ConnectDiscoveryInfoView.as_view(),
        name="oidc-connect-discovery-info",
    ),
    re_path(r"^protocol/openid-connect/userinfo$", oauth2_views.UserInfoView.as_view(), name="user-info"),
    re_path(r"^protocol/openid-connect/auth$", AuthorizationView.as_view(), name="authorize"),
    re_path(r"^protocol/openid-connect/registrations$", RegistrationView.as_view(), name="registrations"),
    re_path(r"^protocol/openid-connect/token$", oauth2_views.TokenView.as_view(), name="token"),
    re_path(r"^protocol/openid-connect/logout$", LogoutView.as_view(), name="logout"),
    re_path(r"^account$", EditUserInfoView.as_view(), name="edit_user_info"),
    re_path(r"^login-actions/action-token$", views.ActionToken.as_view(), name="action-token"),
]
