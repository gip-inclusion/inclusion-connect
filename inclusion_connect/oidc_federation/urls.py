from django.urls import include, re_path

from . import peama


app_name = "oidc_federation"

peama_urlpatterns = [
    re_path(r"^callback/", peama.OIDCAuthenticationCallbackView.as_view(), name="callback"),
    re_path(r"^authenticate/", peama.OIDCAuthenticationRequestView.as_view(), name="init"),
    # There's no logout available on PEAMA
]

urlpatterns = [
    re_path(r"^peama/", include((peama_urlpatterns, "peama"))),
]
