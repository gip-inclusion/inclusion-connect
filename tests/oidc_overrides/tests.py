import datetime

import pytest
from django.contrib.auth import get_user
from django.contrib.sessions.models import Session
from django.test import override_settings
from django.urls import reverse
from django.utils import timezone
from freezegun import freeze_time
from pytest_django.asserts import assertRedirects

from inclusion_connect.users.models import UserApplicationLink
from inclusion_connect.utils.oidc import OIDC_SESSION_KEY
from inclusion_connect.utils.urls import add_url_params
from tests.helpers import (
    OIDC_PARAMS,
    call_logout,
    has_ongoing_sessions,
    oidc_complete_flow,
    parse_response_to_soup,
    token_are_revoked,
)
from tests.oidc_overrides.factories import ApplicationFactory
from tests.users.factories import DEFAULT_PASSWORD, UserFactory


class TestRedirectUris:
    @pytest.mark.parametrize("allow_all", [True, False])
    def test_allow_all_settings(self, allow_all):
        with override_settings(ALLOW_ALL_REDIRECT_URIS=allow_all):
            application = ApplicationFactory(redirect_uris="*")
            assert application.redirect_uri_allowed("http://localhost/callback") is allow_all

    def test_allow_wildcard_at_redirect_uris_end(self):
        application = ApplicationFactory(redirect_uris="http://localhost/*")
        assert application.redirect_uri_allowed("http://localhost/callback")

    def test_no_open_redirect_uri(self):
        application = ApplicationFactory(redirect_uris="http://localhost*")
        assert not application.redirect_uri_allowed("http://localhost/callback")


class TestPostLogoutRedirectUris:
    @pytest.mark.parametrize("allow_all", [True, False])
    def test_allow_all_settings(self, allow_all):
        with override_settings(ALLOW_ALL_REDIRECT_URIS=allow_all):
            application = ApplicationFactory(post_logout_redirect_uris="*")
            assert application.post_logout_redirect_uri_allowed("http://localhost/callback") is allow_all

    def test_allow_wildcard_at_redirect_uris_end(self):
        application = ApplicationFactory(post_logout_redirect_uris="http://localhost/*")
        assert application.post_logout_redirect_uri_allowed("http://localhost/callback")

    def test_no_open_redirect_uri(self):
        application = ApplicationFactory(post_logout_redirect_uris="http://localhost*")
        assert not application.post_logout_redirect_uri_allowed("http://localhost/callback")


class TestLogoutView:
    def test_id_token_hint(self, client):
        """This test simulates a call on logout endpoint with id_hint params"""
        user = UserFactory()
        id_token = oidc_complete_flow(client, user)

        assert get_user(client).is_authenticated is True

        response = call_logout(
            client,
            "get",
            {"id_token_hint": id_token, "post_logout_redirect_uri": "http://callback/"},
        )
        assertRedirects(response, "http://callback/", fetch_redirect_response=False)

        assert get_user(client).is_authenticated is False
        assert has_ongoing_sessions(user) is False
        assert token_are_revoked(user) is True

    def test_expired_token_and_session(self, client):
        """This test simulates a call on logout endpoint with expired token and sessions"""
        user = UserFactory()
        with freeze_time("2023-05-05 14:29:20"):
            id_token = oidc_complete_flow(client, user)
            assert get_user(client).is_authenticated is True

        with freeze_time("2023-05-05 14:59:21"):
            params = {"id_token_hint": id_token, "post_logout_redirect_uri": "http://callback/"}
            response = call_logout(client, "get", params)
            assert response.status_code == 200

            params["allow"] = True
            response = call_logout(client, "post", params)
            assertRedirects(response, "http://callback/", fetch_redirect_response=False)

            assert get_user(client).is_authenticated is False
            assert has_ongoing_sessions(user) is False
            assert token_are_revoked(user) is True

    def test_bad_login_hint(self, client):
        """This test simulates a call on logout endpoint with bad login hint"""
        user = UserFactory()
        oidc_complete_flow(client, user)

        assert get_user(client).is_authenticated is True

        response = call_logout(client, "get", {"id_token_hint": 111})
        assert response.status_code == 400

        assert token_are_revoked(user) is False
        assert get_user(client).is_authenticated is True
        assert has_ongoing_sessions(user) is True


class TestAuthorizeView:
    def test_bad_oidc_params(self, client, snapshot):
        # Application does not exist
        auth_url = reverse("oidc_overrides:authorize")
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        assert response.status_code == 400
        assert str(parse_response_to_soup(response, selector="main")) == snapshot

    def test_not_authenticated(self, client):
        ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
        auth_url = reverse("oidc_overrides:authorize")
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        assertRedirects(response, reverse("accounts:login"))
        assert client.session["next_url"] == auth_complete_url
        assert client.session[OIDC_SESSION_KEY] == OIDC_PARAMS


class TestRegisterView:
    def test_bad_oidc_params(self, client, snapshot):
        # Application does not exist
        auth_url = reverse("oidc_overrides:register")
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        assert response.status_code == 400
        assert str(parse_response_to_soup(response, selector="main")) == snapshot

    def test_not_authenticated(self, client):
        ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
        auth_url = reverse("oidc_overrides:register")
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        assertRedirects(response, reverse("accounts:register"))
        assert client.session["next_url"] == auth_complete_url
        assert client.session[OIDC_SESSION_KEY] == OIDC_PARAMS


class TestActivateView:
    def test_bad_oidc_params(self, client, snapshot):
        auth_url = reverse("oidc_overrides:activate")
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        assert response.status_code == 400
        assert str(parse_response_to_soup(response, selector="main")) == snapshot

    def test_missing_user_info(self, client, snapshot):
        ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
        auth_url = reverse("oidc_overrides:activate")
        # Missing: email, firstname and lastname.
        auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
        response = client.get(auth_complete_url)
        # The user is redirected to the activation view as the oidc parameters are valid
        assertRedirects(response, reverse("accounts:activate"), fetch_redirect_response=False)
        assert client.session["next_url"] == auth_complete_url
        assert client.session[OIDC_SESSION_KEY] == OIDC_PARAMS

        response = client.get(response.url)
        assert response.status_code == 400
        assert str(parse_response_to_soup(response, selector="main")) == snapshot

    def test_not_authenticated(self, client):
        ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
        auth_params = OIDC_PARAMS | {"login_hint": "email", "firstname": "firstname", "lastname": "lastname"}
        auth_url = reverse("oidc_overrides:activate")
        auth_complete_url = add_url_params(auth_url, auth_params)
        response = client.get(auth_complete_url)
        assertRedirects(response, reverse("accounts:activate"))
        assert client.session["next_url"] == auth_complete_url
        assert client.session[OIDC_SESSION_KEY] == auth_params


def test_user_application_link(client):
    application_1 = ApplicationFactory(client_id="ca713487-f4ac-4283-8429-cab7f0386a00")
    application_2 = ApplicationFactory(client_id="05fb7023-ef66-4b24-896e-35f54a6c637f")
    user = UserFactory()
    client.force_login(user)

    def get_user_application_link_values_list():
        return list(
            UserApplicationLink.objects.values_list("application_id", "user_id", "last_login").order_by(
                "application__client_id"
            )
        )

    auth_params_1 = OIDC_PARAMS.copy()
    auth_params_1["client_id"] = application_1.client_id
    auth_url_1 = add_url_params(reverse("oidc_overrides:authorize"), auth_params_1)
    auth_params_2 = OIDC_PARAMS.copy()
    auth_params_2["client_id"] = application_2.client_id
    auth_url_2 = add_url_params(reverse("oidc_overrides:authorize"), auth_params_2)

    assert user.linked_applications.count() == 0

    with freeze_time("2023-04-27 14:06"):
        dt_1 = timezone.now()
        client.get(auth_url_1)

    assert get_user_application_link_values_list() == [(application_1.pk, user.pk, dt_1)]

    with freeze_time("2023-04-27 14:07"):
        dt_2 = timezone.now()
        client.get(auth_url_2)

    assert get_user_application_link_values_list() == [
        (application_2.pk, user.pk, dt_2),
        (application_1.pk, user.pk, dt_1),
    ]

    with freeze_time("2023-04-27 14:08"):
        dt_3 = timezone.now()
        client.get(auth_url_1)

    assert get_user_application_link_values_list() == [
        (application_2.pk, user.pk, dt_2),
        (application_1.pk, user.pk, dt_3),  # last_login was updated
    ]


def test_session_duration(client):
    application_1 = ApplicationFactory()
    application_2 = ApplicationFactory()
    user = UserFactory()

    auth_params = OIDC_PARAMS.copy()
    auth_url = reverse("oidc_overrides:authorize")

    auth_params["client_id"] = application_1.client_id
    auth_complete_url = add_url_params(auth_url, auth_params)
    with freeze_time("2023/05/12 10:39"):
        now = timezone.now()
        response = client.get(auth_complete_url)
        assertRedirects(response, reverse("accounts:login"))
        response = client.post(
            response.url,
            data={
                "email": user.email,
                "password": DEFAULT_PASSWORD,
            },
        )
        assertRedirects(response, auth_complete_url, fetch_redirect_response=False)

    session = Session.objects.get()
    assert session.expire_date == now + datetime.timedelta(minutes=30)

    # 1O minutes later
    auth_params["client_id"] = application_2.client_id
    auth_complete_url = add_url_params(auth_url, auth_params)
    with freeze_time("2023/05/12 10:49"):
        response = client.get(auth_complete_url)
        assert response.status_code == 302
        assert response.url.startswith(OIDC_PARAMS["redirect_uri"])

    # No change in expire_date
    session = Session.objects.get()
    assert session.expire_date == now + datetime.timedelta(minutes=30)
