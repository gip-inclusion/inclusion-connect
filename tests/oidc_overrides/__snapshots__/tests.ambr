# serializer version: 1
# name: TestActivateView.test_bad_oidc_params_400
  '''
  <main class="s-main" id="main" role="main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      <div class="row g-0 bg-white">
                          
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
      <div class="block-center">
          
              <h2>Error: invalid_request</h2>
              <p>Invalid client_id parameter value.</p>
          
      </div>
  
          </div>
      </div>
  
                          
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img alt="" class="img-fluid" loading="lazy" src="/static/img/illustration-ic.svg"/>
      </div>
  
                      </div>
                  </div>
              </section>
  
          </main>
  '''
# ---
# name: TestActivateView.test_missing_user_info
  '''
  <main class="s-main" id="main" role="main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      <div class="row g-0 bg-white">
                          
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
      <div class="block-center">
          
              <h2>Error: invalid_request</h2>
              <p>Missing activation parameters</p>
          
      </div>
  
          </div>
      </div>
  
                          
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img alt="" class="img-fluid" loading="lazy" src="/static/img/illustration-ic.svg"/>
      </div>
  
                      </div>
                  </div>
              </section>
  
          </main>
  '''
# ---
# name: TestAuthorizeView.test_bad_oidc_params_400
  '''
  <main class="s-main" id="main" role="main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      <div class="row g-0 bg-white">
                          
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
      <div class="block-center">
          
              <h2>Error: invalid_request</h2>
              <p>Invalid client_id parameter value.</p>
          
      </div>
  
          </div>
      </div>
  
                          
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img alt="" class="img-fluid" loading="lazy" src="/static/img/illustration-ic.svg"/>
      </div>
  
                      </div>
                  </div>
              </section>
  
          </main>
  '''
# ---
# name: TestRegisterView.test_bad_oidc_params_400
  '''
  <main class="s-main" id="main" role="main">
  
              <section class="s-section pt-lg-3">
                  <div class="container">
                      <div class="row g-0 bg-white">
                          
      <div class="col-12 col-lg-6 bg-lg-white">
          <div class="bg-white p-3 py-lg-7 px-lg-8">
              <div class="container container-messages">
                  
                      
  
  
                  
              </div>
  
              
      <div class="block-center">
          
              <h2>Error: invalid_request</h2>
              <p>Invalid client_id parameter value.</p>
          
      </div>
  
          </div>
      </div>
  
                          
      <div class="d-none d-lg-flex col-lg-6 p-lg-6 justify-content-center align-items-center bg-light">
          <img alt="" class="img-fluid" loading="lazy" src="/static/img/illustration-ic.svg"/>
      </div>
  
                      </div>
                  </div>
              </section>
  
          </main>
  '''
# ---
